daemon/image.go

```diff
@@ -113,12 +121,228 @@ func (daemon *Daemon) GetCachedImage(imgID image.ID, config *containertypes.Conf
+		historyWithSource[i].createdAt = h.Created
+		historyWithSource[i].history = h
+	}
+
+	return historyWithSource
+}
+
 // GetCachedImageOnBuild returns a reference to a cached image whose parent equals `parent`
 // and runconfig equals `cfg`. A cache miss is expected to return an empty ID and a nil error.
-func (daemon *Daemon) GetCachedImageOnBuild(imgID string, cfg *containertypes.Config) (string, error) {
-	cache, err := daemon.GetCachedImage(image.ID(imgID), cfg)
-	if cache == nil || err != nil {
+func (cache *daemonImageCacheForBuild) GetCachedImageOnBuild(imgID string, cfg *containertypes.Config) (string, error) {
+	cachedImage, err := cache.daemon.GetCachedImage(image.ID(imgID), cfg)
+	if err != nil {
 		return "", err
 	}
-	return cache.ID().String(), nil
+	if cachedImage != nil {
+		// We found a cache hit using the old parent image method
+		return cachedImage.ID().String(), nil
+	}
+
+	// We didn't find a cache hit using that method. Explore cacheFrom images for matching history
+	var parentImage *image.Image
+	var parentImageHistory []historyWithSourceT
+	// i.e. not FROM SCRATCH
+	if imgID != "" {
+		parentImage, err = cache.daemon.imageStore.Get(image.ID(imgID))
+		if err != nil {
+			return "", err
+		}
+		parentImageHistory = makeHistoryWithSource(parentImage)
+	}
+
+	// For each thing we are caching from, see if it matches parentImageHistory
+	type matchStruct struct {
+		cacheFrom   string
+		nextLayerID layer.DiffID
+		createdAt   time.Time
+		history     image.History
+	}
+	matches := make([]matchStruct, 0, len(cache.cacheFromImages))
+	for cacheFromName, cacheFromImageHistory := range cache.cacheFromImageHistories {
+
+		if historiesMatch(cacheFromImageHistory, parentImageHistory) &&
+			// Not only do the histories need to match, but cacheFromImageHistory needs to be long enough
+			// that we can peek at the next piece of "history" (which is in the future for parentImage -
+			// that's why we can cache from it!)
+			(len(cacheFromImageHistory) > len(parentImageHistory)) &&
+			// And the command needs to match too.
+			(strings.Join(cfg.Cmd, " ") == cacheFromImageHistory[len(parentImageHistory)].cmd) {
+
+			match := matchStruct{
+				cacheFrom:   cacheFromName,
+				nextLayerID: cacheFromImageHistory[len(parentImageHistory)].resultingLayerID,
+				createdAt:   cacheFromImageHistory[len(parentImageHistory)].createdAt,
+				history:     cacheFromImageHistory[len(parentImageHistory)].history,
+			}
+			matches = append(matches, match)
+		}
+	}
+
+	// Pluck out the newest layer from the potential matches
+	// Pointer so it can be nil
+	var newestMatch *matchStruct
+	for _, match := range matches {
+		if newestMatch == nil || match.createdAt.After(newestMatch.createdAt) {
+			newestMatch = &match
+		}
+	}
+
+	// If we have a match, build an image from it to represent the next state of the build
+	if newestMatch != nil {
+		// TODO: I just copied this from daemon/commit.go. I'm not sure how/if it should be shared.
```

Comment:
Yes. This really is `Commit` but instead of a container rootfs+containerconfig comes from existing image. Maybe combine the common parts or add interface input.